### PR TITLE
feat(core): add `Collection.slice()` method

### PR DESCRIPTION
**Pull request by rubiin**
Adds support for `Collection.slice()`

### CODE REVIEW

File: packages/core/src/entity/Collection.ts [hunk 1]
```diff
@@ -204,6 +204,17 @@ export class Collection<T extends object, O extends object = object> extends Arr
     return super.isEmpty();
   }
 
+
+  /**
+   * Extracts a slice of elements starting at position $start to $end (exclusive) of the Collection.
+   * If $end is null it returns all elements from $start to the end of the Collection.
```
Comment (B4nan) [inline packages/core/src/entity/Collection.ts:210]: comments should be on the base class, and here you should have `@inheritDoc` tag

also there is no `$start` or `$end` parameter, its just `start` and `end`

File: tests/EntityManager.mongo.test.ts [hunk 2]
```diff
@@ -846,6 +846,12 @@ describe('EntityManagerMongo', () => {
     book = (await orm.em.findOne(Book, book._id))!;
     expect(book.tags.count()).toBe(2);
 
+    // slice
+    expect(book.tags.slice()).toEqual(items);
+    expect(book.tags.slice(0,2)).toEqual(items);
+    expect(book.tags.slice(1)).toEqual(items[1]);
+    expect(book.tags.slice(0,1)).toEqual(items[0]);
```
Comment (B4nan) [inline tests/EntityManager.mongo.test.ts:853]: formatting, same for all the other tests

```suggestion
    expect(book.tags.slice(0, 2)).toEqual(items);
    expect(book.tags.slice(1)).toEqual(items[1]);
    expect(book.tags.slice(0, 1)).toEqual(items[0]);
```

and the assertions are most probably wrong, how can you compare an array of items to the item (the second line)? you need to compare arrays with arrays, even if they have a single item in them

**Review by B4nan — CHANGES_REQUESTED** [review]

File: packages/core/src/entity/ArrayCollection.ts [hunk 3]
```diff
@@ -177,6 +177,32 @@ export class ArrayCollection<T extends object, O extends object> {
...
+   */
+
+  slice(start = 0, end?: number): T[]  {
+    let counter = 0;               // counter to keep track of the index of the collection
+    end = end ?? this.items.size; // if end is undefined, set it to the size of the collection
+    const items = new Set<T>();
+
+    for (const item of this.items) {
+
+    if (counter === end) { break; } // break the loop if we reach the end of the slice
```
Comment (B4nan) [inline packages/core/src/entity/ArrayCollection.ts:193]:
```suggestion
    if (counter === end) {
      break;
    }
```

File: packages/core/src/entity/ArrayCollection.ts [hunk 4]
```diff
@@ -177,6 +177,32 @@ export class ArrayCollection<T extends object, O extends object> {
...
+
+    if (counter >= start &&  counter < end) {
```
Comment (B4nan) [inline packages/core/src/entity/ArrayCollection.ts:195]:
```suggestion
    if (counter >= start && counter < end) {
```

we really need to make this fail in eslint, i am sure there is a rule for this

File: packages/core/src/entity/ArrayCollection.ts [hunk 5]
```diff
@@ -177,6 +177,32 @@ export class ArrayCollection<T extends object, O extends object> {
...
+   * Extracts a slice of elements starting at position start to end (exclusive) of the Collection.
+   * If end is null it returns all elements from $start to the end of the Collection.
+   */
+
+  slice(start = 0, end?: number): T[]  {
+    let counter = 0;               // counter to keep track of the index of the collection
+    end = end ?? this.items.size; // if end is undefined, set it to the size of the collection
+    const items = new Set<T>();
+
+    for (const item of this.items) {
```
Comment (B4nan) [inline packages/core/src/entity/ArrayCollection.ts:191]: wrong indent in the rest of this method

File: packages/core/src/entity/Collection.ts [hunk 6]
```diff
@@ -204,6 +204,14 @@ export class Collection<T extends object, O extends object = object> extends Arr
     return super.isEmpty();
   }
 
+  /**
+   * @inheritDoc
+   */
+  slice(start = 0, end?: number): T[]  {
```
Comment (B4nan) [inline packages/core/src/entity/Collection.ts:210]: no need for the default here, its enough to have it in the base method
```suggestion
  slice(start?: number, end?: number): T[]  {
```

File: tests/EntityManager.mongo.test.ts [hunk 7]
```diff
@@ -846,6 +846,12 @@ describe('EntityManagerMongo', () => {
     book = (await orm.em.findOne(Book, book._id))!;
     expect(book.tags.count()).toBe(2);
 
+    // slice
+    expect(book.tags.slice()).toEqual(items);
+    expect(book.tags.slice(0, 2)).toEqual(items);
+    expect(book.tags.slice(1)).toEqual(items[1]);
```
Comment (B4nan) [inline tests/EntityManager.mongo.test.ts:852]: as already said, this is wrong, you cant compare array with the item itself, should be something like following instead:

```suggestion
    expect(book.tags.slice(1)).toEqual([items[1], items[2]]);
```

**Review by B4nan — CHANGES_REQUESTED** [review]

File: packages/core/src/entity/ArrayCollection.ts (hunk 3 above)
Comment (rubiin) [reply packages/core/src/entity/ArrayCollection.ts]: I think we should also setup prettier to run on lint stage

**Comment by rubiin** [comment]
I was thinking instead of adding some minor collection methods in next pr, I will add them in here itself. What do you say

**Comment by B4nan** [comment]
i'd say it should be separate PRs so it results in multiple items in the changelog

and please pay more attention to the comments, you still havent addressed all the missing spaces and wrong indents

**Comment by rubiin** [comment]
I do have editorconfig extension setup and it should take care of the indentation afaik

File: packages/core/src/entity/ArrayCollection.ts [hunk 8]
```diff
@@ -177,6 +177,33 @@ export class ArrayCollection<T extends object, O extends object> {
...
+      if (index >= start && index < end) {
+        items.push(item);
+      }
+
+      index++;
+    }
+
+    return items;
+
+  }
```
Comment (B4nan) [inline packages/core/src/entity/ArrayCollection.ts:205]:
```suggestion
  }
```

File: tests/EntityManager.mysql.test.ts [hunk 9]
```diff
@@ -1188,6 +1188,12 @@ describe('EntityManagerMySql', () => {
     book = (await orm.em.findOne(Book2, book.uuid, { populate: ['tags'] as const }))!;
     expect(book.tags.count()).toBe(3);
 
+    // slice
+    expect(book.tags.slice().length).toBe(3);
+    expect(book.tags.slice(0,3).length).toBe(3);
```
Comment (B4nan) [inline tests/EntityManager.mysql.test.ts:1193]: this is still present all over the changes

**Review by B4nan — CHANGES_REQUESTED** [review]

File: tests/EntityManager.mysql.test.ts (hunk 9 above)
Comment (rubiin) [reply tests/EntityManager.mysql.test.ts:1193]: By this, did you mean the indent

Comment (B4nan) [reply tests/EntityManager.mysql.test.ts:1193]: no, I mean the missing space, I told you about this problem 3 times already :]

Comment (B4nan) [reply tests/EntityManager.mysql.test.ts:1193]: checked the eslint config, and we are missing this:

```
'comma-spacing': 'error',
```

so feel free to add that here to your PR and fix it via eslint, this rule should have a fixer

Comment (rubiin) [reply tests/EntityManager.mysql.test.ts]: Added and that took care of the spacing.

File: .eslintrc.js [hunk 10]
```diff
@@ -148,7 +148,8 @@ module.exports = {
     'space-before-blocks': ['error', 'always'],
     'brace-style': ['error', '1tbs', { allowSingleLine: true }],
     'keyword-spacing': ['error', { before: true, after: true }],
-    'space-in-parens': ['error', 'never']
+    'space-in-parens': ['error', 'never'],
+    'comma-spacing': 'error'
```
Comment (B4nan) [inline .eslintrc.js:152]:
```suggestion
    'comma-spacing': 'error',
```

File: packages/core/src/entity/ArrayCollection.ts [hunk 11]
```diff
@@ -177,6 +177,32 @@ export class ArrayCollection<T extends object, O extends object> {
...
+   * If end is null it returns all elements from start to the end of the Collection.
+   */
+  slice(start = 0, end?: number): T[] {
+    let index = 0;
+    end ??= this.items.size;
+    const items: T[] = [];
+
+    for (const item of this.items) {
+
+      if (index === end) {
```
Comment (B4nan) [inline packages/core/src/entity/ArrayCollection.ts:192]:
```suggestion
      if (index === end) {
```

File: tests/QueryBuilder.test.ts [hunk 12]
```diff
@@ -2942,7 +2942,7 @@ describe('QueryBuilder', () => {
     spy.mockRestore();
   });
 
-  test('limit of 0 limits results to 0',()=>{
+  test('limit of 0 limits results to 0', ()=>{
```
Comment (B4nan) [inline tests/QueryBuilder.test.ts:2945]: hmm this is missing some spaces as well around the arrow, but that must have some eslint rule as well... damn, I though I got it set up well, but apparently I am missing some very basic rules in there.

```suggestion
  test('limit of 0 limits results to 0', () => {
```

File: tests/features/composite-keys/custom-pivot-entity.sqlite.test.ts [hunk 13]
```diff
@@ -197,6 +197,12 @@ describe('custom pivot entity for m:n with additional properties (bidirectional)
     order = (await orm.em.findOne(Order, order.id, { populate: ['products'] as const }))!;
     expect(order.products.count()).toBe(3);
 
+
+    // slice
```
Comment (B4nan) [inline tests/features/composite-keys/custom-pivot-entity.sqlite.test.ts:201]:
```suggestion
    // slice
```

**Review by B4nan — APPROVED** [review]
all right, last few style changes. sorry for not having the linter set up properly, I hate to waste time (yours as well as mine) discussing style changes like this.

thanks!

**Comment by rubiin** [comment]
> all right, last few style changes. sorry for not having the linter set up properly, I hate to waste time (yours as well as mine) discussing style changes like this.
> 
> thanks!

No worries man. With codebase of this size, these are bound to happen.

**Comment by rubiin** [comment]
Btw about the collections test, should I just add tests for a specific driver under `features/Collection.ts` or will the file hold for all drivers

**Comment by B4nan** [comment]
We don't have to test this on all the drivers, in-memory sqlite is enough.

**Pull request merged**